### PR TITLE
Server side code for log scroll

### PR DESCRIPTION
**Pull request by tempoz**
Currently there is nothing to prevent the event logs from being written as chunks (where the frontend currently cannot retrieve them), so I intend to add a config flag to toggle this which defaults to "off" before I merge.

---

**Version bump**: Patch

### CODE REVIEW

File: server/backends/chunkstore/chunkstore.go
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:31]: Can this be removed?

File: server/backends/chunkstore/chunkstore.go
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:96]: Is this slow? Seems like a lot of round trips. Might be something we can optimize away later by saving the last chunk index in the invocation row when we Finalize it

File: server/backends/chunkstore/chunkstore.go
Comment (tylerwilliams) [inline server/backends/chunkstore/chunkstore.go:398]: nit: log.Debugf instead of fmt.Printf (or just remove this)

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tylerwilliams) [inline server/build_event_protocol/build_event_handler/build_event_handler.go:73]: nit: can you define both of these numbers as constants in a const () block at the top of this file? ie:

const (
  // Chunks will be flushed to blobstore when they reach this size.
  defaultChunkSize: 1_000_000 

  // Chunks will also be flushed to blobstore after this much time
  // passes with no new data being written.
  defaultChunkTimeout: 15 * time.Second
)

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tylerwilliams) [inline server/build_event_protocol/build_event_handler/build_event_handler.go:75]: nit: we should probably define a factory function for assembling these remote blobstore "paths" so you don't end up with 2 different paths because of a typo

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tylerwilliams) [inline server/build_event_protocol/build_event_handler/build_event_handler.go:355]: Where do we write to the screen writer / terminal thingy currently? Could we write to the chunkstore in that location too or is there a good reason to have this in a different location?

File: server/eventlog/eventlog.go
Comment (tylerwilliams) [inline server/eventlog/eventlog.go:35]: ubernit: req and rsp

File: server/eventlog/eventlog.go
Comment (tylerwilliams) [inline server/eventlog/eventlog.go:41]: nit: use your factory method here to assemble the remote chunk path

(maybe have that factory method live on chunkstore and be public?)

File: server/eventlog/eventlog.go
Comment (tylerwilliams) [inline server/eventlog/eventlog.go:15]: nit: add a comment about what it does

**Review by tylerwilliams — COMMENTED** [review]
few comments but this looks good overall!

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
(oops, meant to hit request changes)

File: server/backends/chunkstore/chunkstore.go
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:31]: yes

File: server/backends/chunkstore/chunkstore.go
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:398]: I'll just remove it, it's vestigial

File: server/backends/chunkstore/chunkstore.go
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:96]: that makes sense; this definitely could be slow depending on the number of blobs / latency to the blobstore

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:73]: Yes; didn't know about underscores in numeric literals; that's neat

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:75]: good point.

File: server/eventlog/eventlog.go
Comment (tempoz) [reply server/eventlog/eventlog.go:35]: easy enough.

File: server/eventlog/eventlog.go
Comment (tempoz) [reply server/eventlog/eventlog.go:15]: will do.

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:355]: The place it is currently being written is in the [EventParser](https://github.com/buildbuddy-io/buildbuddy/blob/master/server/build_event_protocol/event_parser/event_parser.go#L142-L148), which is also used to look up invocations, which is not a time we want to be writing chunks. I thought this would be better, since even if we decided to conditionally write chunks, it's still a weird side effect to add to the function we use to parse events.

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:73]: huh, those didn't work for me; nvm

Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:73]: OOOOH they work they just don't get syntax highlighted

File: proto/invocation.proto
Comment (tylerwilliams) [inline proto/invocation.proto:84]: protos should be c_style_names

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tylerwilliams) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:367]: SGTM

Since this mutates the event (p.Progress.Std* = ""), does it get called before / after that code or interfere with it?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:367]: It gets called before, which prevents the event log from being stored in the invocation if we are using the chunked logs strategy (by design). Does that seem good to you? Are there alternatives that you might consider preferable?

File: server/backends/chunkstore/chunkstore.go
Comment (tempoz) [reply server/backends/chunkstore/chunkstore.go:95]: Leveraging the invocation table for this now.

File: server/build_event_protocol/build_event_handler/build_event_handler.go
Comment (tylerwilliams) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:367]: (putting this here for completeness) We should continue to write the data in the old-style *as well* as the new style, and then control which style is displayed based on a flag. That way if we need to roll back we can easily do so. Once we are happy that the new-style stuff is 100% awesome, then we can stop writing old-style logs.

Comment (tylerwilliams) [reply server/build_event_protocol/build_event_handler/build_event_handler.go:367]: (also -- rest of this CL LGTM, so once this is fixed i'll approve)

Comment (tempoz) [reply server/build_event_protocol/build_event_handler/build_event_handler.go]: Fixed it! Removed the erasure and ensured the config flag will be pushed over the wire to the client

**Review by tylerwilliams — APPROVED** [review]
LGTM

Before you submit this -- we'll want to add the LastChunkId field to the dev database. Otherwise the apps will attempt to modify the schema themselves, timeout, and get into a restart loop.

**Comment by tempoz** [comment]
Added to dev with
```
ALTER TABLE `Invocations` ADD `last_chunk_id` varchar(255), ALGORITHM=INPLACE, LOCK=NONE;
```

**Pull request merged**
